File: backend/src/utils/performance_monitor.py
"""
Performance Monitor for Backend

Provides performance monitoring utilities for measuring function execution time.
"""

import time
from functools import wraps
from typing import Callable, TypeVar, Any, Dict, List

T = TypeVar('T')


class PerformanceMonitor:
    """Monitor and measure performance of function execution"""
    
    def __init__(self):
        self._metrics: Dict[str, List[float]] = {}
        self._warnings: Dict[str, List[str]] = {}
    
    def measure(self, name: str, fn: Callable[..., T]) -> T:
        """
        Measure execution time of a function
        
        Args:
            name: Name of the metric
            fn: Function to measure
            
        Returns:
            Result of the function
        """
        start_time = time.time()
        result = fn()
        duration = time.time() - start_time
        
        if name not in self._metrics:
            self._metrics[name] = []
        self._metrics[name].append(duration)
        
        # Warning: 1 second threshold
        if duration > 1.0:
            warning = f"Performance warning: {name} took {duration:.2f}s"
            print(warning)
            if name not in self._warnings:
                self._warnings[name] = []
            self._warnings[name].append(warning)
        
        return result
    
    def get_stats(self, name: str) -> Dict[str, Any]:
        """
        Get statistics for a specific metric
        
        Args:
            name: Name of the metric
            
        Returns:
            Dictionary with statistics
        """
        metrics = self._metrics.get(name, [])
        
        if not metrics:
            return {
                'avg': 0,
                'min': 0,
                'max': 0,
                'count': 0
            }
        
        return {
            'avg': sum(metrics) / len(metrics),
            'min': min(metrics),
            'max': max(metrics),
            'count': len(metrics)
        }
    
    def get_all_metrics(self) -> Dict[str, Dict[str, Any]]:
        """
        Get all metrics
        
        Returns:
            Dictionary mapping metric names to statistics
        """
        result = {}
        for name in self._metrics.keys():
            result[name] = self.get_stats(name)
        return result
    
    def clear(self) -> None:
        """Clear all metrics"""
        self._metrics.clear()
        self._warnings.clear()
    
    def get_warnings(self) -> Dict[str, List[str]]:
        """
        Get all warnings
        
        Returns:
            Dictionary mapping metric names to warnings
        """
        return self._warnings.copy()
    
    def has_warnings(self) -> bool:
        """
        Check if there are any warnings
        
        Returns:
            True if there are warnings, False otherwise
        """
        return len(self._warnings) > 0


# Global performance monitor instance
performance_monitor = PerformanceMonitor()


def monitor_performance(name: str):
    """
    Decorator for measuring function performance
    
    Args:
        name: Name of the metric
        
    Returns:
        Decorator function
        
    Example:
        @monitor_performance('calculate_correlation')
        def calculate_correlation(x, y):
            return x * y
    """
    def decorator(func: Callable[..., T]) -> Callable[..., T]:
        @wraps(func)
        def wrapper(*args, **kwargs):
            start_time = time.time()
            result = func(*args, **kwargs)
            duration = time.time() - start_time
            
            if name not in performance_monitor._metrics:
                performance_monitor._metrics[name] = []
            performance_monitor._metrics[name].append(duration)
            
            # Warning: 1 second threshold
            if duration > 1.0:
                warning = f"Performance warning: {name} took {duration:.2f}s"
                print(warning)
                if name not in performance_monitor._warnings:
                    performance_monitor._warnings[name] = []
                performance_monitor._warnings[name].append(warning)
            
            return result
        
        return wrapper
    
    return decorator

File: trading-platform/app/shared/utils/performanceMonitor.ts
/**
 * Performance Monitor
 * 
 * Utility for monitoring and measuring performance in the application.
 */

export interface PerformanceMetric {
  name: string;
  duration: number;
  timestamp: number;
}

export interface PerformanceStats {
  avg: number;
  min: number;
  max: number;
  count: number;
}

export class PerformanceMonitor {
  private static metrics: Map<string, number[]> = new Map();
  private static warnings: Set<string> = new Set();

  /**
   * Measure execution time of a function
   */
  static measure(name: string, fn: () => void): void {
    const start = performance.now();
    fn();
    const duration = performance.now() - start;

    if (!this.metrics.has(name)) {
      this.metrics.set(name, []);
    }
    this.metrics.get(name)!.push(duration);

    // Warning: 100ms threshold
    if (duration > 100) {
      console.warn(`Performance warning: ${name} took ${duration.toFixed(2)}ms`);
      this.warnings.add(name);
    }
  }

  /**
   * Measure execution time of an async function
   */
  static async measureAsync(name: string, fn: () => Promise<void>): Promise<void> {
    const start = performance.now();
    await fn();
    const duration = performance.now() - start;

    if (!this.metrics.has(name)) {
      this.metrics.set(name, []);
    }
    this.metrics.get(name)!.push(duration);

    // Warning: 100ms threshold
    if (duration > 100) {
      console.warn(`Performance warning: ${name} took ${duration.toFixed(2)}ms`);
      this.warnings.add(name);
    }
  }

  /**
   * Get statistics for a specific metric
   */
  static getStats(name: string): PerformanceStats {
    const metrics = this.metrics.get(name) || [];

    if (metrics.length === 0) {
      return { avg: 0, min: 0, max: 0, count: 0 };
    }

    return {
      avg: metrics.reduce((a, b) => a + b, 0) / metrics.length,
      min: Math.min(...metrics),
      max: Math.max(...metrics),
      count: metrics.length
    };
  }

  /**
   * Get all metrics
   */
  static getAllMetrics(): Map<string, PerformanceStats> {
    const result = new Map<string, PerformanceStats>();

    for (const [name, values] of this.metrics.entries()) {
      result.set(name, this.getStats(name));
    }

    return result;
  }

  /**
   * Clear all metrics
   */
  static clear(): void {
    this.metrics.clear();
    this.warnings.clear();
  }

  /**
   * Get all warnings
   */
  static getWarnings(): string[] {
    return Array.from(this.warnings);
  }

  /**
   * Check if there are any warnings
   */
  static hasWarnings(): boolean {
    return this.warnings.size > 0;
  }
}

// Decorator for measuring function performance
export function measurePerformance(name: string) {
  return function (
    target: unknown,
    propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const originalMethod = descriptor.value;

    descriptor.value = function (this: unknown, ...args: unknown[]) {
      const start = performance.now();
      const result = originalMethod.apply(this, args);
      const duration = performance.now() - start;

      if (!PerformanceMonitor['metrics'].has(name)) {
        PerformanceMonitor['metrics'].set(name, []);
      }
      PerformanceMonitor['metrics'].get(name)!.push(duration);

      // Warning: 100ms threshold
      if (duration > 100) {
        console.warn(`Performance warning: ${name} took ${duration.toFixed(2)}ms`);
        PerformanceMonitor['warnings'].add(name);
      }

      return result;
    };

    return descriptor;
  };
}

// Export singleton instance
export const performanceMonitor = PerformanceMonitor;
